src/Compilers/Core/Portable/SourceGeneration/GeneratorTimerExtensions.cs

```diff
@@ -23,40 +23,42 @@ public static RunTimer CreateGeneratorDriverRunTimer(this CodeAnalysisEventSourc
             }
             else
             {
-                return new RunTimer();
+                return new RunTimer(adjustRunTime: null);
             }
         }
 
-        public static RunTimer CreateSingleGeneratorRunTimer(this CodeAnalysisEventSource eventSource, ISourceGenerator generator)
+        public static RunTimer CreateSingleGeneratorRunTimer(this CodeAnalysisEventSource eventSource, ISourceGenerator generator, Func<TimeSpan, TimeSpan> adjustRunTime)
         {
             if (eventSource.IsEnabled(EventLevel.Informational, Keywords.Performance))
             {
                 var id = Guid.NewGuid().ToString();
                 var type = generator.GetGeneratorType();
                 eventSource.StartSingleGeneratorRunTime(type.FullName!, type.Assembly.Location, id);
-                return new RunTimer(t => eventSource.StopSingleGeneratorRunTime(type.FullName!, type.Assembly.Location, t.Ticks, id));
+                return new RunTimer(t => eventSource.StopSingleGeneratorRunTime(type.FullName!, type.Assembly.Location, t.Ticks, id), adjustRunTime);
             }
             else
             {
-                return new RunTimer();
+                return new RunTimer(adjustRunTime);
             }
         }
 
         internal readonly struct RunTimer : IDisposable
         {
             private readonly SharedStopwatch _timer;
             private readonly Action<TimeSpan>? _callback;
+            private readonly Func<TimeSpan, TimeSpan>? _adjustRunTime;
 
-            public TimeSpan Elapsed => _timer.Elapsed;
+            public TimeSpan Elapsed => _adjustRunTime is not null ? _adjustRunTime(_timer.Elapsed) : _timer.Elapsed;
 
-            public RunTimer()
+            public RunTimer(Func<TimeSpan, TimeSpan>? adjustRunTime = null)
```

Comment:
> = null [](http://example.com/codeflow?start=67&length=7)

Consider removing the optional value, since callers will need to specify the value to use this constructor rather than the parameterless constructor.